### PR TITLE
KEP-4210: kubelet: add reason field to image gc metric

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
add image gc reason to metric
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Add "reason" field to image_garbage_collected_total metric, so admins can differentiate images that were collected for reason "age" vs "space"
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4210
```

### CODE REVIEW

**Comment by kannon92** [comment]
/retitle KEP-4210: kubelet: add reason field to image gc metric

File: pkg/kubelet/images/image_gc_manager.go [hunk 1]
```diff
@@ -50,6 +50,14 @@ const instrumentationScope = "k8s.io/kubernetes/pkg/kubelet/images"
 // indexed as imageId-RuntimeHandler
 const imageIndexTupleFormat = "%s,%s"
 
+// ImageGarbageCollectedTotalReason* specify the reason why
```
Comment (kannon92) [inline pkg/kubelet/images/image_gc_manager.go:53]: nit: maybe its worth documenting each one.

File: pkg/kubelet/images/image_gc_manager.go [hunk 2]
```diff
@@ -50,6 +50,14 @@ const instrumentationScope = "k8s.io/kubernetes/pkg/kubelet/images"
...
+// an image was garbage collected, to help operators decide if their
+// workloads are healthy as a result of enabling ImageGCMaxAge feature.
```
Comment (kannon92) [inline pkg/kubelet/images/image_gc_manager.go:55]: I don't really understand this comment. Why would GC impact if workloads are healthy?

**Comment by kannon92** [comment]
/lgtm
non-blocking comments on some documentation in the comments.
/assign @mrunalp @dchen1107

File: pkg/kubelet/images/image_gc_manager.go (hunk 2 above)
Comment (haircommander) [reply pkg/kubelet/images/image_gc_manager.go:55]: the risk is thrashing with GC and a CronJob or something where an image is removed soon before it is set to be run. Documenting the reason will let an operator see if excessive image gc'ing is from this field or just happening generally.

Comment (kannon92) [reply pkg/kubelet/images/image_gc_manager.go:55]: So this metric is only valid if ImageGCMaxAge is enabled? I believe space existed before your feature?

Comment (haircommander) [reply pkg/kubelet/images/image_gc_manager.go:55]: right with the "reason" they'll differentiate the reason between "age" (added in this kep) and "space" (the old reason it used to happen, which still can happen now)

Comment (kannon92) [reply pkg/kubelet/images/image_gc_manager.go:55]: My issue is that this isn't really that clear with just this standalone comment (if I didn't know anything about the KEP).

Comment (kannon92) [reply pkg/kubelet/images/image_gc_manager.go:55]: I would just leave 
// ImageGarbageCollectedTotalReason* specify the reason why
// an image was garbage collected

Comment (haircommander) [reply pkg/kubelet/images/image_gc_manager.go]: fair enough, updated

**Review by saschagrunert — APPROVED** [review]

**Review by mrunalp — APPROVED** [review]

**Pull request merged**